### PR TITLE
Adding main APIObjects

### PR DESCRIPTION
**Pull request by Sigmanificient**
This pull requests add the main API Objects in order to prepare version 0.5.0 and later.

### CODE REVIEW

**Comment by Sigmanificient** [comment]
> Long classes names, Should we remove file name to shorten them?

File: pincer/utils/snowflake.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+class Snowflake(int):
```
Comment (Endercheif) [inline pincer/utils/snowflake.py:25]: change `var` to `param`

**Review by Endercheif — CHANGES_REQUESTED** [review]
Some variables should be documented. If we dont want a description add `#:` after the variable.
Ex:
```py
foo: int = 10 #:
```

Make sure `param` is used instead of `var`
Document errors with `:raises ERRORNAME: DESCRIPTION`
Inline code in reST should be \`\`some code\`\` instead of  \`some code\`.

**Comment by Endercheif** [comment]
The [codacy issue](https://app.codacy.com/gh/Pincer-org/Pincer/file/60703061619/issues/source?bid=25184624&fileBranchId=25320428#l37) is not an issue due to `__future__.annotations`

File: pincer/objects/embed.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+    :param proxy_icon_url: A proxied url of the author icon
+    """
+    icon_url: APINullable[str] = MISSING
+    name: APINullable[str] = MISSING
+    proxy_icon_url: APINullable[str] = MISSING
+    url: APINullable[str] = MISSING
+
+    def __post_init__(self):
+        if _field_size(self.name) > 256:
+            raise EmbedFieldError.from_desc("Author name", 256, len(self.name))
```
[the remark this answers is not in the document]
Comment (Lunarmagpie) [reply pincer/objects/embed.py:84]: so it should be like:
```:raises EmbedFieldError: Something about fields```
or
```:raises: Something about fields```
?

[the remark this answers is not in the document]
Comment (Endercheif) [reply pincer/objects/embed.py:84]: `:raises EmbedFieldError: Something about fields`

File: pincer/objects/channel.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+class ChannelType(Enum):
+    GUILD_TEXT = 0
+    DM = 1
+    GUILD_VOICE = 2
+    GROUP_DM = 3
+    GUILD_CATEGORY = 4
+    GUILD_NEWS = 5
+    GUILD_STORE = 6
+
+    if GatewayConfig.version >= 9:
```
[the remark this answers is not in the document]
Comment (Arthurdw) [reply pincer/objects/channel.py:50]: Yes, this works.

File: pincer/objects/__init__.py [hunk 4]
```diff
@@ -21,3 +21,19 @@
...
+from .embed import Embed
+from .emoji import Emoji
+from .guild import Guild
+from .invite import Invite
+from .member import Member
+from .role import Role
+from .sticker import Sticker
+from .user import User
+
+__all__ = (
```
Comment (Sigmanificient) [inline pincer/objects/__init__.py:36]: A lot of objects are missing in the `__all__`

File: pincer/objects/embed.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+            proxy_url=proxy_url,
+            width=width
+        )
+
+    def set_provider(
+        self,
+        name: APINullable[str] = MISSING,
+        url: APINullable[str] = MISSING
+    ) -> Embed:
+        # TODO: Provide general documentation of what a provider is.
```
Comment (Arthurdw) [inline pincer/objects/embed.py:412]: Provide general documentation of what a provider is.

File: pincer/objects/invite.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+from pincer.objects.channel import Channel
+from pincer.objects.guild import Guild
+from pincer.objects.user import User
+from pincer.objects.member import Member
+from pincer.objects.application import Application
+from pincer.utils.constants import APINullable, MISSING
+from pincer.utils.timestamp import Timestamp
+
+
+class InviteTargetType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/invite.py:41]: Write documentation for this

File: pincer/objects/application.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+from pincer.utils.api_object import APIObject
+
+from pincer.objects.user import User
+from pincer.utils.constants import APINullable, MISSING
+from pincer.utils.snowflake import Snowflake
+
+
+@dataclass
+class Application(APIObject):
+    """
```
Comment (Arthurdw) [inline pincer/objects/application.py:38]: Write general description

File: pincer/objects/application_command.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+from dataclasses import dataclass
+from enum import Enum
+from typing import List, Union
+
+from pincer.utils.api_object import APIObject
+from pincer.utils.constants import MISSING, APINullable
+from pincer.utils.snowflake import Snowflake
+
+
+class ApplicationCommandType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/application_command.py:35]: Write general and specific docs for properties and enum itself

File: pincer/objects/application_command.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    CHAT_INPUT = 1
+    USER = 2
+    MESSAGE = 3
+
+
+class ApplicationCommandOptionType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/application_command.py:41]: Write general and specific docs for properties and enum itself

File: pincer/objects/message.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+from pincer.objects.reaction import Reaction
+from pincer.objects.role import Role
+from pincer.objects.sticker import StickerItem
+from pincer.objects.user import User
+from pincer.utils.api_object import APIObject
+from pincer.utils.constants import MISSING, APINullable
+from pincer.utils.snowflake import Snowflake
+from pincer.utils.timestamp import Timestamp
+
+class MessageActivityType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/message.py:47]: Write docs

File: pincer/objects/message.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    CROSSPOSTED = 1 << 0
+    IS_CROSSPOST = 1 << 1
+    SUPPRESS_EMBEDS = 1 << 2
+    SOURCE_MESSAGE_DELETED = 1 << 3
+    URGENT = 1 << 4
+    HAS_THREAD = 1 << 5
+    EPHEMERAL = 1 << 6
+    LOADING = 1 << 7
+
+class MessageType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/message.py:94]: Write docs

File: pincer/objects/role.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+from dataclasses import dataclass
+
+from pincer.utils.api_object import APIObject
+from pincer.utils.constants import MISSING, APINullable
+from pincer.utils.snowflake import Snowflake
+
+
+@dataclass
+class RoleTags(APIObject):
+    """
```
Comment (Arthurdw) [inline pincer/objects/role.py:35]: Write general documentation

File: pincer/objects/role.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        whether this is the guild's premium subscriber role
+    """
+    bot_id: APINullable[Snowflake] = MISSING
+    integration_id: APINullable[Snowflake] = MISSING
+    premium_subscriber: APINullable[bool] = MISSING
+
+
+@dataclass
+class Role(APIObject):
+    """
```
Comment (Arthurdw) [inline pincer/objects/role.py:52]: Write general documentation

File: pincer/objects/stage.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import Enum
+
+from pincer.utils.snowflake import Snowflake
+from pincer.utils.api_object import APIObject
+
+
+class PrivacyLevel(Enum):
```
Comment (Arthurdw) [inline pincer/objects/stage.py:33]: Write general docs

File: pincer/objects/sticker.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+from enum import Enum
+from typing import List, Optional
+
+from pincer.objects.user import User
+from pincer.utils.api_object import APIObject
+from pincer.utils.constants import MISSING, APINullable
+from pincer.utils.snowflake import Snowflake
+
+
+class StickerType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/sticker.py:36]: Write general docs

File: pincer/objects/sticker.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    STANDARD = 1
+    GUILD = 2
+
+
+class StickerFormatType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/sticker.py:41]: Write general docs

File: pincer/objects/sticker.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    PNG = 1
+    APNG = 2
+    LOTTIE = 3
+
+
+@dataclass
+class Sticker(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/sticker.py:48]: Write general docs

File: pincer/objects/sticker.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+
+    available: APINullable[bool] = MISSING
+    guild_id: APINullable[Snowflake] = MISSING
+    pack_id: APINullable[Snowflake] = MISSING
+    sort_value: APINullable[int] = MISSING
+    user: APINullable[User] = MISSING
+
+
+@dataclass
+class StickerItem(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/sticker.py:102]: Write general docs

File: pincer/objects/sticker.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        type of sticker format
+    """
+
+    id: Snowflake
+    name: str
+    format_type: StickerFormatType
+
+
+@dataclass
+class StickerPack(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/sticker.py:120]: Write general docs

File: pincer/objects/user.py [hunk 20]
```diff
@@ -38,32 +40,73 @@ class PremiumTypes(Enum):
...
-    verified: Optional[bool] = False #: Whether the email on this account has been verified
-    avatar_url: Optional[str] = None #: The url to the user's avatar
-    mfa_enabled: Optional[bool] = False #: Whether the user has two factor enabled on their account
-    premium_type: Optional[PremiumTypes] = PremiumTypes.NONE #: The type of Nitro subscription on a user's account
-
-    @classmethod
-    def from_dict(cls, data: Data[str, Union[str, bool, int]]) -> User:
-        # TODO: Write documentation
-        return cls(**data)
+class User(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/user.py:43]: Write general docs

File: pincer/objects/webhook.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+from pincer.objects.channel import Channel
+from pincer.objects.guild import Guild
+from pincer.objects.user import User
+from pincer.utils.api_object import APIObject
+from pincer.utils.constants import APINullable, MISSING
+from pincer.utils.snowflake import Snowflake
+
+
+class WebhookType(Enum):
```
Comment (Arthurdw) [inline pincer/objects/webhook.py:38]: Write general docs

File: pincer/objects/welcome_screen.py [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+
+from dataclasses import dataclass
+from typing import Optional
+
+from pincer.utils.api_object import APIObject
+from pincer.utils.snowflake import Snowflake
+
+
+@dataclass
+class WelcomeScreenChannel(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/welcome_screen.py:34]: Write general docs

File: pincer/objects/welcome_screen.py [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+
+    channel_id: Snowflake
+    description: str
+
+    emoji_id: Optional[int] = None
+    emoji_name: Optional[str] = None
+
+
+@dataclass
+class WelcomeScreen(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/welcome_screen.py:59]: Write general docs

File: pincer/utils/api_object.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+        )
+    else:
+        return copy.deepcopy(obj)
+
+
+@dataclass
+class APIObject:
+
+    @classmethod
+    def from_dict(cls, data: Data[str, Union[str, bool, int]]) -> APIObject:
```
Comment (Arthurdw) [inline pincer/utils/api_object.py:80]: Write docs

File: pincer/utils/api_object.py [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+        # TODO: Write documentation
+        return cls(**data)
+
+    def to_dict(self) -> Dict:
```
Comment (Arthurdw) [inline pincer/utils/api_object.py:84]: Write docs

File: pincer/utils/api_object.py [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+                _asdict_ignore_none(k, dict_factory),
+                _asdict_ignore_none(v, dict_factory)
+            ) for k, v in obj.items()
+        )
+    else:
+        return copy.deepcopy(obj)
+
+
+@dataclass
+class APIObject:
```
Comment (Arthurdw) [inline pincer/utils/api_object.py:78]: Write general documentation

File: pincer/utils/snowflake.py (hunk 1 above)
Comment (Arthurdw) [reply pincer/utils/snowflake.py:25]: Resolved

File: pincer/utils/snowflake.py [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+    :var process_id:
+        Internal process ID
+
+    :var increment:
+        For every ID that is generated on that process,
+        this number is incremented
+    """
+
+    @property
+    def timestamp(self) -> int:
```
Comment (Arthurdw) [inline pincer/utils/snowflake.py:54]: Write docs

File: pincer/utils/snowflake.py [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+        return self >> 22
+
+    @property
+    def worker_id(self) -> int:
```
Comment (Arthurdw) [inline pincer/utils/snowflake.py:58]: Write docs

File: pincer/utils/snowflake.py [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+        return (self >> 17) % 16
+
+    @property
+    def process_id(self) -> int:
```
Comment (Arthurdw) [inline pincer/utils/snowflake.py:62]: Write docs

File: pincer/utils/snowflake.py [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+        return (self >> 12) % 16
+
+    @property
+    def increment(self) -> int:
```
Comment (Arthurdw) [inline pincer/utils/snowflake.py:66]: Write docs

File: pincer/utils/timestamp.py [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+
+from __future__ import annotations
+from datetime import datetime
+from typing import Optional, TypeVar, Union
+
+DISCORD_EPOCH = 1420070400
+TS = TypeVar("TS", str, datetime, float, int)
+
+
+class Timestamp:
```
Comment (Arthurdw) [inline pincer/utils/timestamp.py:33]: Write general docs

File: pincer/utils/timestamp.py [hunk 32]
```diff
@@ -0,0 +1,96 @@
...
+
+    def __init__(self, time: Optional[TS] = None):
+        self.__time = Timestamp.parse(time)
+        self.__epoch = Timestamp.to_epoch(self.__time)
+        self.date, self.time = str(self).split()
+
+    @staticmethod
+    def to_epoch(time: datetime) -> int:
```
Comment (Arthurdw) [inline pincer/utils/timestamp.py:41]: Write docs

File: pincer/utils/timestamp.py [hunk 33]
```diff
@@ -0,0 +1,96 @@
...
+        return int(time.timestamp() * 1000)
+
+    @staticmethod
+    def string_to_datetime(string: str) -> datetime:
```
Comment (Arthurdw) [inline pincer/utils/timestamp.py:45]: Write docs

File: pincer/utils/timestamp.py [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+        return datetime.fromisoformat(string)
+
+    @staticmethod
+    def epoch_to_datetime(epoch: Union[int, float]) -> datetime:
```
Comment (Arthurdw) [inline pincer/utils/timestamp.py:49]: Write docs

File: pincer/utils/timestamp.py [hunk 35]
```diff
@@ -0,0 +1,96 @@
...
+        return datetime.utcfromtimestamp(epoch)
+
+    @staticmethod
+    def parse(time: Optional[TS]) -> datetime:
```
Comment (Arthurdw) [inline pincer/utils/timestamp.py:53]: Write docs

File: pincer/objects/guild.py [hunk 36]
```diff
@@ -0,0 +1,272 @@
...
+    threads: APINullable[List[Channel]] = MISSING
+    # Guilds are considered available unless otherwise specified
+    unavailable: APINullable[bool] = False
+    voice_states: APINullable[bool] = MISSING
+    widget_enabled: APINullable[bool] = MISSING
+    widget_channel_id: APINullable[Optional[Snowflake]] = MISSING
+    welcome_screen: APINullable[WelcomeScreen] = MISSING
+
+    @classmethod
+    def from_dict(cls, data) -> Guild:
```
Comment (Arthurdw) [inline pincer/objects/guild.py:266]: Write docs

File: pincer/objects/guild_member.py [hunk 37]
```diff
@@ -0,0 +1,79 @@
...
+
+from pincer.utils.api_object import APIObject
+from pincer.objects.user import User
+from pincer.utils.constants import APINullable, MISSING
+from pincer.utils.snowflake import Snowflake
+from pincer.utils.timestamp import Timestamp
+
+
+@dataclass
+class GuildMember(APIObject):
```
Comment (Arthurdw) [inline pincer/objects/guild_member.py:37]: Write general documentation

File: pincer/objects/guild_member.py [hunk 38]
```diff
@@ -0,0 +1,79 @@
...
+
+    :param joined_at:
+        when the user joined the guild
+
+    :param mute:
+        whether the user is muted in voice channels
+
+    :param roles:
+        array of role object ids
+
```
Comment (Arthurdw) [inline pincer/objects/guild_member.py:50]: Remove unnecessary space

File: pincer/objects/embed.py [hunk 39]
```diff
@@ -0,0 +1,550 @@
...
+
+    @classmethod
+    def from_desc(cls, _type: str, max_size: int, cur_size: int):
+        return cls(
+            f"{_type} can have a maximum length of {max_size}."
+            f" (Current size: {cur_size})"
+        )
+
+
+class InvalidUrlError(ValueError):
```
Comment (Arthurdw) [inline pincer/objects/embed.py:45]: Write documentation for this

Comment (Arthurdw) [reply pincer/objects/embed.py:45]: & move to exceptions class

File: pincer/objects/guild_member.py (hunk 38 above)
Comment (Arthurdw) [reply pincer/objects/guild_member.py:52]: Resolved

File: pincer/objects/message.py [hunk 40]
```diff
@@ -0,0 +1,262 @@
...
+    channel_id: Snowflake
+    author: User
+    content: str
+    timestamp: Timestamp
+    edited_timestamp: Optional[Timestamp]
+    tts: bool
+    mention_everyone: bool
+    mentions: List[User]
+    mention_roles: List[Role]
+    mention_channels: List[ChannelMention] # FIXME
```
Comment (Arthurdw) [inline pincer/objects/message.py:242]: Resolve fixme

File: pincer/objects/message.py (hunk 11 above)
Comment (Arthurdw) [reply pincer/objects/message.py:106]: Resolved

File: pincer/objects/message.py (hunk 10 above)
Comment (Arthurdw) [reply pincer/objects/message.py:50]: Resolved

File: pincer/objects/user.py (hunk 20 above)
Comment (Arthurdw) [reply pincer/objects/user.py:45]: Resolved

File: pincer/objects/invite.py (hunk 6 above)
Comment (Arthurdw) [reply pincer/objects/invite.py:41]: Resolved.

File: pincer/objects/role.py (hunk 12 above)
Comment (Arthurdw) [reply pincer/objects/role.py:35]: Resolved

File: pincer/objects/role.py (hunk 13 above)
Comment (Arthurdw) [reply pincer/objects/role.py:56]: Resolved

File: pincer/objects/stage.py (hunk 14 above)
Comment (Arthurdw) [reply pincer/objects/stage.py:33]: Resolved

File: pincer/objects/sticker.py (hunk 15 above)
Comment (Arthurdw) [reply pincer/objects/sticker.py:36]: Resolved

File: pincer/objects/sticker.py (hunk 17 above)
Comment (Arthurdw) [reply pincer/objects/sticker.py:69]: Resolved

File: pincer/objects/sticker.py (hunk 16 above)
Comment (Arthurdw) [reply pincer/objects/sticker.py:50]: Resolved

File: pincer/objects/sticker.py (hunk 18 above)
Comment (Arthurdw) [reply pincer/objects/sticker.py:125]: Resolved

File: pincer/objects/sticker.py (hunk 19 above)
Comment (Arthurdw) [reply pincer/objects/sticker.py:146]: Resolved

File: pincer/objects/webhook.py (hunk 21 above)
Comment (Arthurdw) [reply pincer/objects/webhook.py:38]: Resolved

File: pincer/objects/__init__.py (hunk 4 above)
Comment (Arthurdw) [reply pincer/objects/__init__.py:58]: Resolved by @trag1c

File: pincer/utils/api_object.py (hunk 26 above)
Comment (Arthurdw) [reply pincer/utils/api_object.py:77]: Resolved.

File: pincer/utils/api_object.py (hunk 25 above)
Comment (Arthurdw) [reply pincer/utils/api_object.py:92]: Resolved

File: pincer/objects/welcome_screen.py (hunk 22 above)
Comment (Arthurdw) [reply pincer/objects/welcome_screen.py:34]: Resolved

File: pincer/objects/welcome_screen.py (hunk 23 above)
Comment (Arthurdw) [reply pincer/objects/welcome_screen.py:61]: Resolved

File: pincer/utils/snowflake.py [hunk 41]
```diff
@@ -0,0 +1,67 @@
...
+    identifiable descriptors (IDs).
+
+    These IDs are guaranteed to be unique across all of Discord,
+    except in some unique scenarios in which child objects
+    share their parent's ID.
+
+    Because Snowflake IDs are up to 64 bits in size (e.g. a uint64),
+    they are always returned as strings in the HTTP API
+    to prevent integer overflows in some languages.
+
```
Comment (Arthurdw) [inline pincer/utils/snowflake.py:37]: Move docs to property

File: pincer/utils/timestamp.py (hunk 31 above)
Comment (Arthurdw) [reply pincer/utils/timestamp.py:33]: Resolved

File: pincer/utils/timestamp.py (hunk 33 above)
Comment (Arthurdw) [reply pincer/utils/timestamp.py:53]: Resolved

File: pincer/utils/timestamp.py (hunk 34 above)
Comment (Arthurdw) [reply pincer/utils/timestamp.py:63]: Resolved

File: pincer/utils/timestamp.py (hunk 35 above)
Comment (Arthurdw) [reply pincer/utils/timestamp.py]: Resolved

File: pincer/utils/timestamp.py (hunk 32 above)
Comment (Arthurdw) [reply pincer/utils/timestamp.py:43]: Resolved

**Review by Arthurdw — APPROVED** [review]
Fully reviewed, ready for main lib

File: pincer/objects/role.py [hunk 42]
```diff
@@ -0,0 +1,96 @@
...
+    color: int
+    hoist: bool
+    id: Snowflake
+    managed: bool
+    mentionable: bool
+    name: str
+    permissions: str
+    position: int
+
+    tags: APINullable[...] = MISSING
```
Comment (trag1c) [inline pincer/objects/role.py:96]: Add RoleTags

File: pincer/objects/user.py [hunk 43]
```diff
@@ -25,54 +25,105 @@
...
     @property
     def premium(self) -> PremiumTypes:
-        # TODO: Write documentation
+        """
+        The user their premium type in a usable enum.
+        """
         return PremiumTypes(self.premium_type)
 
     @property
     def user(self) -> str:
```
Comment (trag1c) [inline pincer/objects/user.py:123]: Not quite sure if `User.user` is a good name

File: pincer/utils/timestamp.py [hunk 44]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        Convert an epoch to a datetime object.
+
+        :param epoch:
+            The epoch to convert to a datetime object.
+        """
+        return datetime.utcfromtimestamp(epoch)
+
+    @staticmethod
+    def parse(time: Optional[TS] = None) -> datetime:
```
Comment (trag1c) [inline pincer/utils/timestamp.py:73]: Add support for type `datetime` as an argument

**Review by trag1c — CHANGES_REQUESTED** [review]

File: pincer/objects/user.py (hunk 43 above)
Comment (Arthurdw) [reply pincer/objects/user.py]: Removed the property and replaced with str fmt.

File: pincer/utils/timestamp.py (hunk 44 above)
Comment (Arthurdw) [reply pincer/utils/timestamp.py:73]: U fixed this urself :)

**Review by Arthurdw — APPROVED** [review]
Ready for merge

**Pull request merged**